Defer omnifunc only the first time; testing vimscript through vimrunner-python

vheon opened the pull request:
I wanted to get it out and get some real feedback about it. This PR supersedes #1770 since it enables testing the vimscript that handle this deferring through [vimrunner-python](/andri-ch/vimrunner-python). The cons of this:

- No Windows at the moment.
- No python 2.6

The cons are actually limitation of `vimrunner-python`. @micbou do you have news about this? I know you were working on it, just checking ;)

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/1811)
<!-- Reviewable:end -->

micbou commented:
> The cons are actually limitation of vimrunner-python. @micbou do you have news about this? I know you were working on it, just checking ;)

I was able to add python 2.6 support and to make it work on Windows. Check [my fork](https://github.com/micbou/vimrunner-python). I added Travis and AppVeyor CIs. See the runs [here](https://travis-ci.org/micbou/vimrunner-python) and [here](https://ci.appveyor.com/project/micbou/vimrunner-python/build/job/p0umfrk4d9jwub1y). OS X runs are failing because `x-terminal-emulator` is not available and the failed run on AppVeyor is not a big issue. I stopped working on it because I am considering a rewrite from scratch.

Anyway, I don't think using a testing framework for Vim is urgent and we could merge your PR without the Vim tests.

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1811#-:-K4cf4Ypk_eLwzTIqKJe:1149686587)*
<!-- Sent from Reviewable.io -->

vheon commented:
I had considered to write a vim driver directly in the tests (It doesn't seem that complicated and in the end is really small piece of code) but then I wanted first to try if the approach was valuable. I didn't want to merge vimscript code without reliable tests :confused: What we could do instead is merge this testing only on mac and linux with python 2.7 (some tests  is better then no test IMHO) and then we could replace vimrunner-python with our in-house implementation cross platform and backward compatible; this way we would not merge things blindly and would be always covered. What do you think?

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1811#-:-K4dmJFtkJsTpzcnCdpR:1898804822)*
<!-- Sent from Reviewable.io -->

puremourning commented on travis/travis_install.linux.sh:
```diff
@@ -1,3 +1,31 @@
+  if [[ -d /tmp/vim/.git ]]; then
+    pushd /tmp/vim
+    git fetch
+    if ! git diff --exit-code --quiet ..origin/master; then
```
Are we deliberately checking out master? The question is whether we want to catch upstream changes in vim that break our code or if we want To catch changes in our code which break our code. I think the latter, so I think we should check out a particular commit or tag of vim and build that. We can update it regularly if we like, but we don't want flakiness caused by checking out vim head?

vheon commented:
Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[travis/travis_install.linux.sh, line 9 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1811#-K4Z4fNzsfBbzhpNZK_K-r1-9:-K5uP9PrrnF0pTe96G5K:931876523) ([raw file](https://github.com/valloric/youcompleteme/blob/5760a208f7cd719f3b120d34bced7881e41d3d24/travis/travis_install.linux.sh#L9)):*
We can actually checkout vim 7.3.598 which is the minimum version we support.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1811)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[travis/travis_install.linux.sh, line 9 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1811#-K4Z4fNzsfBbzhpNZK_K-r1-9:-K5wE7tamWoqdUqKbYCm:-214508745) ([raw file](https://github.com/valloric/youcompleteme/blob/5760a208f7cd719f3b120d34bced7881e41d3d24/travis/travis_install.linux.sh#L9)):*
No non-pinned checkouts please, they lead to tons of problems.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1811)*
<!-- Sent from Reviewable.io -->

Valloric commented:
What's the state of this PR? Is it blocked on something?

---

Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1811#-:-K5wERh82EkjY8r4Veg9:-1320294258)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[travis/travis_install.linux.sh, line 9 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1811#-K4Z4fNzsfBbzhpNZK_K-r1-9:-K5yZl3BiltPJFNFqghn:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/5760a208f7cd719f3b120d34bced7881e41d3d24/travis/travis_install.linux.sh#L9)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1811)*
<!-- Sent from Reviewable.io -->

vheon commented:
I've addressed the comment regarding the vim version. As for the state of the PR: I would like to merge this like it is right now using pyvimrunner since even if it doesn't run on python 2.6 and it doesn't run on windows it still gives us a bit of coverage and then in a separate PR we can replace pyvimrunner. As @micbou (and I agree with him) said pyvimrunner is small and simple enough to rewrite it and make it cross platform and make it run on all version of python.

---

Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1811#-:-K5y_2C-o43lXhXUTwTY:-1426186159)*
<!-- Sent from Reviewable.io -->

vheon commented:
Apparently pin the version of `vim` creates some problem. I will investigate.

---

Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1811#-:-K5ybINK9wgFJ7BPh3dc:835604972)*
<!-- Sent from Reviewable.io -->

homu commented:
:umbrella: The latest upstream changes (presumably #1848) made this pull request unmergeable. Please resolve the merge conflicts.

micbou commented:
Why don't we merge #1770 and open a new issue to add Vim tests? It seems dumb to add tests now only for rewriting them later. I am also confident that #1770 is properly working. It has been thoroughly tested by @vheon, @puremourning, and myself.

vheon commented:
@micbou is a reasonable approach, so is fine to me.

Valloric commented:
@micbou Fine by me as well.

micbou commented:
I created [the issue](https://github.com/Valloric/YouCompleteMe/issues/1868).

vheon commented:
The fix is already merged and issue #1868 will track the vim testing. Closing.
